drivers/dma/dma_gd32.c

```diff
@@ -0,0 +1,625 @@
+	gd32_dma_circulation_disable(cfg->reg, channel);
+#if defined(GD32F4XX)
+	gd32_dma_channel_subperipheral_select(cfg->reg, channel, dma_cfg->linked_channel);
+#endif
+
+	data->channels[channel].callback = dma_cfg->dma_callback;
+	data->channels[channel].user_data = dma_cfg->user_data;
+	data->channels[channel].direction = dma_cfg->channel_direction;
+
+	return 0;
+}
+
+static int dma_gd32_reload(const struct device *dev, uint32_t ch, uint32_t src, uint32_t dst,
+			   size_t size)
+{
+	const struct dma_gd32_config *cfg = dev->config;
+	struct dma_gd32_data *data = dev->data;
+
+	if (ch >= cfg->channels) {
+		LOG_ERR("reload channel must be < %d (%d)", cfg->channels, ch);
+		return -EINVAL;
+	}
+
+	if (data->channels[ch].busy) {
+		return -EBUSY;
+	}
+
+	gd32_dma_channel_disable(cfg->reg, ch);
+
+	gd32_dma_transfer_number_config(cfg->reg, ch, size);
+
+	switch (data->channels[ch].direction) {
+	case MEMORY_TO_MEMORY:
+	case PERIPHERAL_TO_MEMORY:
+		gd32_dma_memory_address_config(cfg->reg, ch, dst);
+		gd32_dma_periph_address_config(cfg->reg, ch, src);
+		break;
+	case MEMORY_TO_PERIPHERAL:
+		gd32_dma_memory_address_config(cfg->reg, ch, src);
+		gd32_dma_periph_address_config(cfg->reg, ch, dst);
+		break;
+	}
+
+	gd32_dma_channel_enable(cfg->reg, ch);
+
+	return 0;
+}
+
+static int dma_gd32_start(const struct device *dev, uint32_t ch)
+{
+	const struct dma_gd32_config *cfg = dev->config;
+	struct dma_gd32_data *data = dev->data;
+
+	if (ch >= cfg->channels) {
+		LOG_ERR("start channel must be < %d (%d)", cfg->channels, ch);
+		return -EINVAL;
+	}
+
+	gd32_dma_interrupt_enable(cfg->reg, ch, DMA_FLAG_FTF | GD32_DMA_FLAG_ERRORS);
+	gd32_dma_channel_enable(cfg->reg, ch);
+	data->channels[ch].busy = true;
+
+	return 0;
+}
+
+static int dma_gd32_stop(const struct device *dev, uint32_t ch)
+{
+	const struct dma_gd32_config *cfg = dev->config;
+	struct dma_gd32_data *data = dev->data;
+
+	if (ch >= cfg->channels) {
+		LOG_ERR("stop channel must be < %d (%d)", cfg->channels, ch);
+		return -EINVAL;
+	}
+
+	gd32_dma_interrupt_disable(cfg->reg, ch, DMA_FLAG_FTF | GD32_DMA_FLAG_ERRORS);
+	gd32_dma_interrupt_flag_clear(cfg->reg, ch, DMA_FLAG_FTF | GD32_DMA_FLAG_ERRORS);
+	gd32_dma_channel_disable(cfg->reg, ch);
+	data->channels[ch].busy = false;
+
+	return 0;
+}
+
+static int dma_gd32_get_status(const struct device *dev, uint32_t ch, struct dma_status *stat)
+{
+	const struct dma_gd32_config *cfg = dev->config;
+	struct dma_gd32_data *data = dev->data;
+
+	if (ch >= cfg->channels) {
+		LOG_ERR("channel must be < %d (%d)", cfg->channels, ch);
+		return -EINVAL;
+	}
+
+	stat->pending_length = gd32_dma_transfer_number_get(cfg->reg, ch);
+	stat->dir = data->channels[ch].direction;
+	stat->busy = data->channels[ch].busy;
+
+	return 0;
+}
+
+static bool dma_gd32_api_chan_filter(const struct device *dev, int ch, void *filter_param)
+{
+	uint32_t filter = *((uint32_t *)filter_param);
```

Comment:
Should check `filter_param` is `NULL`.

Reply:
This is required, since dma loop_transfer test `filter_param` is `NULL`.

Reply:
Add null-check.

Reply:
Should be something like below, check filter_param before use.

Thanks for the whole work for gd32 dma. After this one, gd32f450i_eval and gd32f350r_eval pssed the dma loop_transfer test.

```
static bool dma_gd32_api_chan_filter(const struct device *dev, int ch, void *filter_param)
{
	uint32_t filter;

	if (!filter_param) {
		return false;
	}

	filter = *((uint32_t *)filter_param);

      /* .... */
}
```

Reply:
Thanks for verify with gd32f450i_eval and gd32f350r_eval.
I'm happy with the PR work well in these boards.

And also thanks for notice my bug!
It already fixed.

Reply:
`After this one`, means there still have a problem. Pelase consider to use my reference codes, thanks.

Reply:
It is my careless.
Thank you for pointing out.

Fixed and pushed.